### PR TITLE
openjdk: Fix build under WSL 1

### PR DESCRIPTION
**Pull request by eddieqiii**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

The bottled version of openjdk does not work on (my) WSL 1 (#21293), and when building from source the [default behaviour](http://openjdk.java.net/groups/build/doc/building.html) of `./configure` on WSL is to build for a Windows target. This fails when loading the boot jdk since it's expecting a Windows jdk:
```
...
checking if packaged modules are kept... enabled, default
checking for version string... 15.0.2+7
configure: Found potential Boot JDK using configure arguments
configure: Potential Boot JDK found at /tmp/openjdk-20210319-3620-16p05mm/jdk15u-jdk-15.0.2-ga/boot-jdk did not contain bin/java; ignoring
configure: error: The path given by --with-boot-jdk does not contain a valid Boot JDK
configure exiting with result code 1
```

This PR adds a check for WSL and sets the autoconf `--build` and `--host` flags to build for Linux, on both aarch64 and x86_64 (since WSL is supported on both). Notes:
- The rdp/os library does not contain a way to check for WSL ([PR](https://github.com/rdp/os/pull/33/commits/27359223e6ef662139f8601d0f97548ffb5a1dba)) thus the manual uname check.
- `--openjdk-target` is preferred to setting the autoconf flags, however it still configures for Windows when this is set to linux.

### CODE REVIEW

**Comment by danielnachun** [comment]
In your previous issue, I took a look at that error and I noticed something odd - while `adoptopenjdk` was using the host linker `/lib64/ld-linux-x86-64.so.2`, `openjdk` was using the brewed linker `/home/linuxbrew/.linuxbrew/lib/ld.so`.  This is very unusual, and I suspect that ELF error you were getting is happening because the `openjdk` binaries are using the brewed linker and host `glibc` shared libraries (`libc.so.6`, etc.).  

Can you install the `openjdk` bottle in WSL1 and post the output of `ldd /home/linuxbrew/.linuxbrew/opt/openjdk/libexec/bin/javac`?  And also let me know if the `glibc` formula is installed.

I'm not sure we'll be able to accept this PR because it will never be tested in CI.  Generally speaking we prefer to try to fix our bottles whenever possible.

**Comment by eddieqiii** [comment]
It makes sense that you should try to fix the bottles — I was questioning whether a fix for building on WSL belonged in the formula, but decided to fix and submit a PR anyways instead of a duplicate issue. I don't have the Homebrew glibc formula installed on my machine. Here's the output of `ldd` after installing bottled `openjdk`:
```
$ ldd /home/linuxbrew/.linuxbrew/opt/openjdk/libexec/bin/javac
        linux-vdso.so.1 (0x00007fffccee8000)
        libjli.so => /home/linuxbrew/.linuxbrew/opt/openjdk/libexec/bin/../lib/libjli.so (0x00007fc3fd8a0000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc3fd6a0000)
        libz.so.1 => /home/linuxbrew/.linuxbrew/lib/libz.so.1 (0x00007fc3fd480000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc3fd470000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc3fd44d000)
        /home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007fc3fdcf0000)
```
It looks like Homebrew's linker is just a symlink to the host linker on my install since I don't have brew's glibc. The original issue I referenced wasn't mine, so I'll put my `brew config` below in case it helps:
<details>
<summary>
brew config
</summary>

```
HOMEBREW_VERSION: 3.0.7-36-g1b61d5a
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1b61d5a5639d7cf76a1fc3faf02c87f0aef9e64c
Last commit: 10 hours ago
Core tap ORIGIN: https://github.com/Homebrew/linuxbrew-core
Core tap HEAD: 19d738e5ecf1e44e6512a98b3cabec573414b6cb
Core tap last commit: 10 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.3 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: 16-core 64-bit zen
Clang: 10.0 build (parse error)
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /usr/bin/curl
Kernel: Linux 4.4.0-19041-Microsoft x86_64 GNU/Linux
OS: Ubuntu 20.04.1 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.3.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@5: N/A
xorg: N/A
```
</details>

**Comment by danielnachun** [comment]
Thanks for the followup info, this rules out my previous explanation.  I will see if I can dig into this on my Windows machine in WSL 1 as I am very curious as to what would cause this type of error.

**Comment by black-snow** [comment]
Probably related

> ==> ./configure --disable-warnings-as-errors --with-boot-jdk-jvmargs=-Duser.home=/home/xxx/.cache/Homebrew/java_cache --with-boot-jdk=/tmp/openjdk-20210531-14179-16vt5cn/jdk16u-jdk-16.0.1-ga/boot-jdk --with-debug-level=release --
Last 15 lines from /home/xxx/.cache/Homebrew/Logs/openjdk/01.configure:
configure: Found Visual Studio installation at /mnt/c/progra~2/micros~3/2019/BuildTools using well-known name
configure: Warning: None of vc/bin/amd64/vcvars64.bat vc/bin/x86_amd64/vcvarsx86_amd64.bat             vc/auxiliary/build/vcvarsx86_amd64.bat vc/auxiliary/build/vcvars64.bat were found, Visual Studio installation not recognized. Ignoring
configure: Found Visual Studio installation at /mnt/c/progra~2/micros~3/2017/BuildTools using well-known name
configure: Found Microsoft Visual Studio 2017
configure: Trying to extract Visual Studio environment variables for x86_64
configure: using /mnt/c/progra~2/micros~3/2017/BuildTools/vc/auxiliary/build/vcvarsx86_amd64.bat
configure: Setting extracted environment variables for x86_64
checking that Visual Studio variables have been correctly extracted... ok
configure: Will use user supplied compiler CC=gcc-11
checking for gcc-11... /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/gcc-11
checking resolved symbolic links for CC... no symlink
configure: The C compiler (located as /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/gcc-11) does not seem to be the required microsoft compiler.
configure: The result from running it was: "/home/linuxbrew/.linuxbrew/bin/ld: /home/linuxbrew/.linuxbrew/Cellar/gcc/11.1.0_1/lib/gcc/11/gcc/x86_64-pc-linux-gnu/11.1.0/crt1.o: in function `_start':"
configure: error: A microsoft compiler is required. Try setting --with-tools-dir.
configure exiting with result code 1

**Comment by magneticflux-** [comment]
I ran in to the same issue as @black-snow, and I think this patch will fix it. From reading through the `configure` script in `openjdk`, I believe it detects WSL as "Windows" and sets MSVC as the default toolchain and Windows as the default target. Setting the `--build` and `--host` flags manually would prevent this detection and allow it to build for Linux properly.

File: Formula/openjdk.rb [hunk 1]
```diff
@@ -78,6 +78,8 @@ def install
       )
     end
 
+    is_wsl = `uname -r`.include?("Microsoft")
```
Comment (magneticflux-) [inline Formula/openjdk.rb:81]: On WSL2, `uname -r` outputs `4.19.104-microsoft-standard` and the code checks for the string `Microsoft` case-sensitively AFAIK.

**Comment by eddieqiii** [comment]
Yep, this is the build problem that this PR aims to fix, and affects both WSL 1 and 2. It seems that for now openjdk is building from source for everyone which is why this is getting a little more attention (maybe the bottles haven't been built yet?). If this is the case, maybe it's worth adding this check to the formula?

The underlying problem is still that the _bottle_ couldn't run on WSL 1, and as far as I know we're unaware why this is the case. The previous bottled openjdk ran fine on WSL 2 (for me) so it's still a mystery on that front.

**Comment by danielnachun** [comment]
I think this PR will be fine but needs some changes:

1) We need to remove the `is_wsl` line and use `unless OS.mac?` instead. The flags we are adding are fine to add on all Linux installations.  Presumably before they were being added implicitly when building in CI, but setting them explicitly will make sure OpenJDK targets Linux regardless of what `uname` returns.

2) The check for ARM64 needs to be removed as well.  We don't currently support ARM on Linux because there is no CI to test it with, and we don't want to add code to formulae that can't be tested in CI.  If you want to provide ARM64 support, the best way to do this is to make a custom tap from this formula that adds back those checks.

Lastly, the reason OpenJDK 16 is currently being built from source is because it is failing to build in CI.  Hopefully this failure is just because we need to build with a newer GCC, but it could be more complex.  

I haven't had a chance to examine the bottle more closely on WSL 1 but it's on my to do list.  We don't like to have bottles that don't work on some platforms we're supposed to support so hopefully this can be fixed.

**Comment by eddieqiii** [comment]
Noted re the ARM64 change. Regarding the uname check: this was originally an intentional change of mine since the [OpenJDK docs](http://openjdk.java.net/groups/build/doc/building.html) recommend using `--openjdk-target` over setting the autoconf flags individually (under Cross-Compiling: Specifying the Target Platform), and I recall a warning output at the configure step that was to this effect too. However, setting `openjdk-target` still caused the script to configure with the microsoft toolchain. 

Since it's not recommended to set the autoconf flags manually, I thought it'd be best to restrict this to WSL only and keep the bottle build flags untouched. To my knowledge, this shouldn't materially affect anything at this point in time, and it would allow this path to be tested in CI, so if you think it's okay I'll go ahead and make those changes after merging in from upstream.

This should fix building on WSL 1 & 2. I'm also planning to look into why the bottled version won't run on v1, although like many others I've migrated to v2 for the time being. Since v2 is more resource intensive and a fair few are staying on v1, I reckon it's still an issue worth looking into.

**Comment by danielnachun** [comment]
I've enabled the CI run on this.  On the most up to date documentation for building OpenJDK (https://github.com/openjdk/jdk/blob/master/doc/building.md), it actually explicitly says to set these flags to build for Linux in WSL:

>It's possible to build both Windows and Linux binaries from WSL. To build Windows binaries, you must use a Windows boot JDK (located in a Windows-accessible directory). To build Linux binaries, you must use a Linux boot JDK. The default behavior is to build for Windows. To build for Linux, pass `--build=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu` to configure.

So this is a feature not a bug 🙃.

**Comment by danielnachun** [comment]
I just tested the OpenJDK 16 bottle under WSL 1, and it works!  Interestingly, OpenJDK 8 also works fine but OpenJDK 11 does not.  This indicates that some issue was introduced between versions 8 and 11, which was then fixed for 16 (or maybe earlier).  

I also found this comment which actually proposed a fix that worked for OpenJDK 11: https://github.com/microsoft/WSL/issues/5875#issuecomment-688881367.  I wanted to see if I could change the loader with `patchelf` to be `/lib64/ld-linux-x86-64.so.2` instead of `/home/linuxbrew/.linuxbrew/lib/ld.so` and then I hit this error with `patchelf`, identified by @sjackman: https://github.com/NixOS/patchelf/issues/255.  This could mean that there is some sort of alignment issue with the binaries.  This may be fixable by adding `-mstackrealign` to the GCC flags, but I'm not sure.  

It's possible the alignment issue is unrelated, but I did find that I could change the loader with `patchelf` for OpenJDK 8 and OpenJDK 16 with no errors, so it's a suspicious coincidence to the say the least.

**Comment by black-snow** [comment]
Thanks for your efforts guys <3

**Comment by MikeMcQuaid** [comment]
Closing, sorry. If this PR is still relevant: please reopen on homebrew-core.

**Pull request closed without merging**